pkg/broker/util.go

```diff
@@ -290,33 +289,87 @@ func extractProperties(params []apb.ParameterDescriptor) (map[string]*schema.Sch
 			Type:        t,
 		}
 
-		// we can NOT set values on the Schema object if we want to be
-		// omitempty. Setting maxlength to 0 is NOT the same as omitting it.
-		// 0 is a worthless value for Maxlength so we will not set it
-		if pd.Maxlength > 0 {
-			properties[k].MaxLength = schema.Integer{Val: pd.Maxlength, Initialized: true}
-		}
+		setStringValidators(pd, properties[k])
+		setNumberValidators(pd, properties[k])
+		setEnum(pd, properties[k])
+	}
 
-		// do not set the regexp if it does not compile
-		if pd.Pattern != "" {
-			patternRegex, err = regexp.Compile(pd.Pattern)
-			properties[k].Pattern = patternRegex
+	return properties, nil
+}
 
-			if err != nil {
-				fmt.Printf("Invalid pattern: %s", err.Error())
-			}
-		}
+func setStringValidators(pd apb.ParameterDescriptor, prop *schema.Schema) {
+	if prop.Type[0] != schema.StringType {
+		return
+	}
 
-		// setup enums
-		if len(pd.Enum) > 0 {
-			properties[k].Enum = make([]interface{}, len(pd.Enum))
-			for i, v := range pd.Enum {
-				properties[k].Enum[i] = v
-			}
+	// we can NOT set values on the Schema object if we want to be
+	// omitempty. Setting maxlength to 0 is NOT the same as omitting it.
+	// 0 is a worthless value for DeprecatedMaxlength so we will not set it
+
+	// maxlength
+	if pd.DeprecatedMaxlength > 0 {
+		prop.MaxLength = schema.Integer{Val: pd.DeprecatedMaxlength, Initialized: true}
+	}
+
+	// max_length overrides maxlength
+	if pd.MaxLength > 0 {
+		prop.MaxLength = schema.Integer{Val: pd.MaxLength, Initialized: true}
+	}
+	// min_length
+	if pd.MinLength > 0 {
+		prop.MinLength = schema.Integer{Val: pd.MinLength, Initialized: true}
+	}
+
+	// do not set the regexp if it does not compile
+	if pd.Pattern != "" {
+		patternRegex, err := regexp.Compile(pd.Pattern)
+		prop.Pattern = patternRegex
+
+		if err != nil {
+			fmt.Printf("Invalid pattern: %s", err.Error())
 		}
 	}
+}
 
-	return properties, nil
+func setNumberValidators(pd apb.ParameterDescriptor, prop *schema.Schema) {
+	if prop.Type[0] != schema.NumberType && prop.Type[0] != schema.IntegerType {
+		return
+	}
+
+	// since 0 is not useful as a value for multipleOf,
+	// we can use it as a float64 and not worry about nil
+	if pd.MultipleOf > 0 {
+		prop.MultipleOf = schema.Number{Val: pd.MultipleOf, Initialized: true}
+	}
+
+	// since 0 is a valid value for maximum, minimum, exclusiveMaximum, and exclusiveMinimum,
+	// we have to allow for empty.
+	if max, ok := pd.Maximum.(float64); ok {
+		prop.Maximum = schema.Number{Val: max, Initialized: true}
+	}
+	if min, ok := pd.Minimum.(float64); ok {
+		prop.Minimum = schema.Number{Val: min, Initialized: true}
+	}
+
```

Comment:
In the JSON schema [validations documentation](https://datatracker.ietf.org/doc/draft-handrews-json-schema-validation/), `exclusiveMinimum` and `exclusiveMaximum` MUST be a number.  However go-jsschema and the serialized values it sends uses boolean values for `exclusiveMaximum` and `exclusiveMinimum` are boolean and use the numbers in `maximum` and `minimum`.  

I followed with the docs and  used `exclusive_maximum`/`exclusive_minimum` as a number for the APB spec, but if there are strong opinions that I should use boolean `exclusive_maximum`/`exclusive_minimum` instead, I'd like to hear reasoning.  I was divided myself.

Reply:
I agree with following the docs. `exclusive_maximum` representing the max number seems intuitive to me.